Добавила решение задачи

Nigma-Ks opened the pull request.

Nigma-Ks committed:
Добавила решение задачи

ArtyomKopan commented on Homework1/ComparisonMatrixCountingSpeeds.cs:
```diff
@@ -0,0 +1,68 @@
+
+public static class ComparisonMatrixCountingSpeeds
+{
+    public static (double, double)[] Experiments(int sizeOfData, int amountOfTestsInSample) //size of data, mat ozgidanie, dispersiia
```
комментарии транслитом? Ты серьёзно?))

ArtyomKopan commented on Homework1/ComparisonMatrixCountingSpeeds.cs:
```diff
@@ -0,0 +1,68 @@
+        for (int i = 0; i < time.Count; i++)
+        {
+            double difference = time[i] - mathematicalExpectation;
+            sumOfSquareDifference += difference * difference;
```
можно написать короче:
`var sumOfSquareDifference = Enumerable.Range(0, time.Count).Sum(i => (time[i] - mathematicalExpectation) * (time[i] - mathematicalExpectation));`

ArtyomKopan commented on Homework1/Matrix.cs:
```diff
@@ -0,0 +1,174 @@
+    public int[,] matrixOfNumbers { get; private set; }
+    public int Height { get; private set; } = 0;
+    public int Width { get; private set; } = 0;
+    public Matrix(string filePath)
```
лучше передавать в конструктор сразу ширину, высоту и данные. Так положено по правилам инкапсуляции. Вдруг пользователь захочет вводить из консоли?

ArtyomKopan commented on Homework1/Matrix.cs:
```diff
@@ -0,0 +1,174 @@
+        var matrix3Numbers = new int[matrix1.Height, matrix2.Width];
+
+        var matrixMyltiplyingThreads = new Thread[4];
+        for (byte i = 0; i < matrixMyltiplyingThreads.Length; i++)
```
а зачем тут `byte`? Мы же не на Си пишем :)

ArtyomKopan commented on Homework1/Matrix.cs:
```diff
@@ -0,0 +1,174 @@
+        }
+        var matrix3Numbers = new int[matrix1.Height, matrix2.Width];
+
+        var matrixMyltiplyingThreads = new Thread[4];
```
а что, если программа будет запускаться на сервере со 128 процессорами? 4 потока -- это как-то маловато. Лучше написать `var matrixMyltiplyingThreads = new Thread[Environment.ProcessorCount];`

ArtyomKopan commented on Homework1/Matrix.cs:
```diff
@@ -0,0 +1,174 @@
+        for (byte i = 0; i < matrixMyltiplyingThreads.Length; i++)
+        {
+            int numberOfStartElement = i;
+            uint amountOfElementsInMatrix = (uint)matrix1.Height * (uint)matrix2.Width;
```
uint -- тоже слишком низкоуровнево. Не стоит заморачиваться такими вещами, пиши `var amountOfElementsInMatrix = ...` и всё

yurii-litvinov replied:
Всё так, хотя и некритично. Лишние аннотации типов в целом не мешают

ArtyomKopan commented on Homework1/Matrix.cs:
```diff
@@ -0,0 +1,174 @@
+                    matrix3Numbers[currentLine, currentColumn] = 0;
+                    for (int k = 0; k < matrix1.Width; k++)
+                    {
+                        matrix3Numbers[currentLine, currentColumn] += matrix1.matrixOfNumbers[currentLine, k] * matrix2.matrixOfNumbers[k, currentColumn];
```
можно короче написать:
`matrix3Numbers[currentLine, currentColumn] = Enumerable.Range(0, matrix1.Width).Sum(k => matrix1.matrixOfNumbers[currentLine, k] * matrix2.matrixOfNumbers[k, currentColumn]);`

ArtyomKopan commented on Homework1/Matrix.cs:
```diff
@@ -0,0 +1,174 @@
+            uint amountOfElementsInMatrix = (uint)matrix1.Height * (uint)matrix2.Width;
+            matrixMyltiplyingThreads[i] = new Thread(() =>
+            {
+                uint currentElement = (uint)numberOfStartElement;
```
можно сделать проще: 
`var rowsPerThread = Math.Ceiling(matrix2.Height / Environment.ProcessorCount);` и умножать по стольку строк в каждом потоке

ArtyomKopan commented on Homework1/Matrix.cs:
```diff
@@ -0,0 +1,174 @@
+            });
+        }
+
+        foreach (var tread in matrixMyltiplyingThreads)
```
по-английски пишется **thread**.
_tread_ -- это "шагать"

ArtyomKopan commented on Homework1/Program.cs:
было бы неплохо добавить возможность вводить матрицы из консоли

ArtyomKopan commented on Homework1/results.txt:
кажется, у тебя что-то не то с результатами. Как при большой матрице параллельное умножение может работать медленнее последовательного?

ArtyomKopan reviewed (commented):
Пока что 1 балл из 4

Nigma-Ks committed:
- транслит, заменила цикл более короткой записью, количество потоков по количеству ядер

Nigma-Ks committed:
исправила опечатку

Nigma-Ks committed:
1. Исправила функцию, печатающую матрицу, вывод итоговой матрицы в Program
2. Изменила способ счета потоками
3. Обновила результаты экспериментов в табличке results

Nigma-Ks committed:
форматирование

Nigma-Ks committed:
Убрала цикл

Nigma-Ks committed:
Добавила комментарии, namespace'ы

yurii-litvinov commented on Homework1.Tests/MatrixTests.cs:
```diff
@@ -0,0 +1,68 @@
+
+public class Tests
+{
+    Matrix matrix1;
+    Matrix matrix2;
```
```suggestion
    private Matrix matrix1;
    private Matrix matrix2;
```
В .NET по традиции всегда явно пишут модификаторы видимости, даже если умолчания устраивают

yurii-litvinov commented on Homework1.Tests/MatrixTests.cs:
```diff
@@ -0,0 +1,68 @@
+    [Test]
+    public void ExceptionsTest()
+    {
+        Assert.Throws<ArgumentNullException>(() => new Matrix(null));
```
Тут nullability-анализ ругается, если хотите null, напишите `null!`. Надо, чтобы компилировалось без предупреждений.

yurii-litvinov commented on Homework1/ComparisonMatrixCountingSpeeds.cs:
```diff
@@ -0,0 +1,65 @@
+
+using System.Diagnostics;
+
+public static class ComparisonMatrixCountingSpeeds
```
Надо комментарии ко всем типам и всем public-членам

yurii-litvinov commented on Homework1/ComparisonMatrixCountingSpeeds.cs:
```diff
@@ -0,0 +1,65 @@
+        return new (double, double)[] { (mathematicalExpectationByThreads.Value, dispersionByThreads), (mathematicalExpectationSequential.Value, dispersionSequential) };
+    }
+
+    private static double DispersionCalculator(List<long> time, double mathematicalExpectation)
```
Методы должны называться глаголами в повелительной форме

yurii-litvinov commented on Homework1/ComparisonMatrixCountingSpeeds.cs:
```diff
@@ -0,0 +1,65 @@
+        double? mathematicalExpectationSequential = timeSequential.Average();
+        double dispersionByThreads = DispersionCalculator(timeByThreads, mathematicalExpectationByThreads.Value);
+        double dispersionSequential = DispersionCalculator(timeSequential, mathematicalExpectationSequential.Value);
+        return new (double, double)[] { (mathematicalExpectationByThreads.Value, dispersionByThreads), (mathematicalExpectationSequential.Value, dispersionSequential) };
```
Вместо массива кортежей, где всегда должно быть два элемента, сделали бы специальный struct. Явные именованные поля лучше соглашений

yurii-litvinov commented on Homework1/ComparisonMatrixCountingSpeeds.cs:
```diff
@@ -0,0 +1,65 @@
+
+    private static int[,] CreateSquareArrayWithRandomValues(int size)
+    {
+        var random = new Random();
```
Random в принципе должен быть на систему один (во-первых потому, что больше одного не нужно, во-вторых, потому что он инициализируется временем, так что если два объекта Random создать одновременно, они просто одинаковые значения будут выдавать, в-третьих, сборщику мусора их ещё удалять). Так что сделайте его статическим полем.

yurii-litvinov commented on Homework1/Matrix.cs:
```diff
@@ -0,0 +1,180 @@
+        Height = matrixInLines.Length;
+        if (Height < 1)
+        {
+            throw new ArgumentException("There is no numbers in file");
```
```suggestion
            throw new ArgumentException("There are no numbers in file");
```

yurii-litvinov commented on Homework1/Matrix.cs:
```diff
@@ -0,0 +1,180 @@
+                }
+                else
+                {
+                    throw new ArgumentException("Not inly numbers in file");
```
```suggestion
                    throw new ArgumentException("Not only numbers in file");
```

yurii-litvinov commented on Homework1/Matrix.cs:
```diff
@@ -0,0 +1,180 @@
+    /// <summary>
+    /// return matrix in string
+    /// </summary>
+    public string PrintedMatrix()
```
```suggestion
    public string ToString()
```

yurii-litvinov commented on Homework1/Matrix.cs:
```diff
@@ -0,0 +1,180 @@
+        {
+            for (int j = 0; j < Width; j++)
+            {
+                matrixInString += Convert.ToString(MatrixOfNumbers[i, j]) + ' ';
```
Никогда не конкатенируйте строки в цикле, это же постоянное копирование строк и аллокации. Используйте StringBuilder

yurii-litvinov commented on Homework1/Matrix.cs:
```diff
@@ -0,0 +1,180 @@
+    /// </summary>
+    public static Matrix SequentialMultiplyingMatrixes(Matrix matrix1, Matrix matrix2)
+    {
+        if (matrix1 == null)
+        {
+            throw new ArgumentNullException(nameof(matrix1));
+        }
```
```suggestion
        ArgumentNullException.ThrowIfNull(matrix1);
```

yurii-litvinov commented on Homework1/Matrix.cs:
```diff
@@ -0,0 +1,180 @@
+        }
+        var matrix3Numbers = new int[matrix1.Height, matrix2.Width];
+
+        int amountOfThreads = Environment.ProcessorCount;
+        var matrixMyltiplyingThreads = new Thread[amountOfThreads];
```
Так может получиться больше потоков, чем нужно. В матрице может быть меньше Environment.ProcessorCount строк

yurii-litvinov commented on Homework1/Program.cs:
```diff
@@ -0,0 +1,41 @@
+FileStream fs = File.Create(matrix3FilePath);
+if (!File.Exists(matrix3FilePath))
+{
+    throw new ArgumentException("Файл не был создан");
```
В main нет смысла кидать исключения. потому что их даже в теории некому ловить. Выводите ошибку на консоль и всё.

yurii-litvinov commented on Homework1/results.txt:
```diff
@@ -0,0 +1,20 @@
+                 Потоки:  Последовательно:
+Размер матрицы:  200*200
+Мат ожидание:    231     150
+Дисперсия:       1026    135
```
Дисперсия не нужна, нужно среднеквадратичное отклонение
